### PR TITLE
Use clang's effective llvm triple for IR generation

### PR DESCRIPTION
**Pull request by aschwaighofer**
Instead of using the target that was passed to the driver. Use the target from
the clang importer that might have been changed by clang (i.e armv7 to thumbv7
on darwin)

rdar://32599805

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/IRGenModule.cpp [hunk 1]
```diff
@@ -127,11 +127,16 @@ IRGenModule::IRGenModule(IRGenerator &irgen,
       ClangCodeGen(createClangCodeGenerator(Context, LLVMContext, irgen.Opts,
                                             ModuleName)),
       Module(*ClangCodeGen->GetModule()), LLVMContext(Module.getContext()),
-      DataLayout(target->createDataLayout()), Triple(Context.LangOpts.Target),
+      DataLayout(target->createDataLayout()),
+      Triple(static_cast<ClangImporter *>(
+                 &*irgen.SIL.getASTContext().getClangModuleLoader())
+                 ->getTargetInfo()
+                 .getTargetOpts()
+                 .Triple),
```
Comment (jrose-apple) [inline lib/IRGen/IRGenModule.cpp]: This is pretty terrible. Can we just pass it in instead? Or wrap it up in IRGenerator somewhere?

File: test/IRGen/arm_to_thumb_darwin.sil [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// RUN: %swift -target armv7-apple-ios7 %s -gnone -emit-ir -o - | %FileCheck %s -check-prefix=IOS
+// RUN: %swift -target armv7k-apple-watchos2 %s -gnone -emit-ir -o - | %FileCheck %s -check-prefix=WATCHOS
+
+// REQUIRES: CODEGENERATOR=ARM
```
Comment (jrose-apple) [inline test/IRGen/arm_to_thumb_darwin.sil]: This isn't sufficient; the test also can't be run when there's no standard library for iOS or watchOS built.

**Review by jrose-apple — COMMENTED** [review]
Glad you could track this down! Sorry if I wasn't much help.

File: lib/IRGen/IRGen.cpp [hunk 3]
```diff
@@ -482,11 +483,10 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
 
 std::unique_ptr<llvm::TargetMachine>
 swift::createTargetMachine(IRGenOptions &Opts, ASTContext &Ctx) {
-  const llvm::Triple &Triple = Ctx.LangOpts.Target;
   std::string Error;
-  const Target *Target = TargetRegistry::lookupTarget(Triple.str(), Error);
+  const Target *Target = TargetRegistry::lookupTarget(Ctx.LangOpts.Target.str(), Error);
```
Comment (aschwaighofer) [inline lib/IRGen/IRGen.cpp]: Will change.

File: lib/IRGen/IRGenModule.cpp (hunk 1 above)
Comment (aschwaighofer) [inline lib/IRGen/IRGenModule.cpp]: Will change.

File: test/IRGen/arm_to_thumb_darwin.sil (hunk 2 above)
Comment (aschwaighofer) [inline test/IRGen/arm_to_thumb_darwin.sil]: This works (builtin_word.sil also does this and I tested locally with an x86_64 build only). There is no import of the standard library.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: test/IRGen/arm_to_thumb_darwin.sil (hunk 2 above)
Comment (jrose-apple) [inline test/IRGen/arm_to_thumb_darwin.sil]: Oh, I missed that it was `%swift` rather than `%target-build-swift`. Sorry for the noise.

File: test/IRGen/osx-targets.swift [hunk 4]
```diff
@@ -5,7 +5,7 @@
 // REQUIRES: OS=macosx
 
 // CHECK: target triple = "x86_64-apple-macosx10.
-// CHECK-SPECIFIC: target triple = "x86_64-apple-macosx10.12"
+// CHECK-SPECIFIC: target triple = "x86_64-apple-macosx10.12
```
Comment (jrose-apple) [inline test/IRGen/osx-targets.swift]: Why this change? Is there a ".0" sometimes?

**Review by jrose-apple — COMMENTED** [review]
Thanks! One last comment.

File: test/IRGen/osx-targets.swift (hunk 4 above)
Comment (aschwaighofer) [inline test/IRGen/osx-targets.swift]: Yes

Comment (jrose-apple) [inline test/IRGen/osx-targets.swift]: Can you check for that specifically instead of allowing arbitrary patch versions, then?

**Comment by aschwaighofer** [comment]
@swift-ci Please test
